### ISSUE TITLE
SEH support for Windows

### ISSUE THREAD

**Issue by DanyDollaro** [c3lang/c3c#1452]
To provide a comprehensive support for Windows driver development, it would be necessary to make it possible to use the SEH from C3, since it is obligatory during calls of certain kernel functions such as [ProbeForRead](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-probeforread#remarks), an [usage example](https://github.com/microsoft/Windows-driver-samples/blob/b3af8c8f9bd508f54075da2f2516b31d05cd52c8/general/ioctl/wdm/sys/sioctl.c#L409) in C

**Comment by lerno** [comment]
Maybe this isn't useful enough given that there are workarounds we discovered?

**Comment by DanyDollaro** [comment]
Hello Lerno, exactly what workarounds are you talking about?

**Comment by lerno** [comment]
Copy paste from Discord:
---
So compiling this in C:
```c
bool call_with_seh(void* context, CallFn call_fn, ExceptFn except_fn)
{
  try
  {
    call_fn(context);
    return true;
  }
  except(EXCEPTION_EXECUTE_HANDLER)
  {
    except_fn(context);
    return false;
  }
}
```
Then from C3:
```c
def SehFn = fn void(void*);
extern fn bool call_with_seh(void* context, SehFn call_fn, SehFn except_fn);
```

**Comment by DanyDollaro** [comment]
This is really an adoptable solution, I did a little research on how you could use SEH in Rust since it's a very renowned language in windows driver development but currently hasn't caught on yet, and I found this library https://github.com/sonodima/microseh, which actually uses a C file in which it [calls a SEH protected stub](https://github.com/sonodima/microseh/blob/master/src/stub.c), in C3 we could adopt the same technique without having to adopt a special construct in the language

**Comment by lerno** [comment]
Yes, as I think of this, I could actually make a library (.c3l) and put it in vendor which you would include and then on Windows it compiles the C and makes the extension available automatically.

**Comment by lerno** [comment]
Would you be able to test that though? I can write it, but not test it.

**Comment by DanyDollaro** [comment]
Certainly, it will be to see the library as soon as it comes out, for that matter I'm very interested
